### ISSUE TITLE
Scala 2.12 serialization of Props

### ISSUE THREAD

**Issue by patriknw** [akka/akka#22041]
WIth Scala 2.12 the recommendation in http://doc.akka.io/docs/akka/2.4/scala/actors.html#Recommended_Practices doesn't work for remote deployment because Function is no longer serializable. It's translated to a lambda, which is not serializable. E.g. error message:

```
[JVM-2] [ERROR] [12/19/2016 08:42:49.917] [NewRemoteActorSpec-akka.remote.default-remote-dispatcher-14] [akka.remote.Remoting] akka.remote.NewRemoteActorSpec$$$Lambda$459/1116094714
[JVM-2] java.lang.ClassNotFoundException: akka.remote.NewRemoteActorSpec$$$Lambda$459/1116094714
[JVM-2] 	at java.lang.Class.forName0(Native Method)
[JVM-2] 	at java.lang.Class.forName(Class.java:348)
[JVM-2] 	at akka.actor.ReflectiveDynamicAccess.$anonfun$getClassFor$1(ReflectiveDynamicAccess.scala:21)
[JVM-2] 	at akka.actor.ReflectiveDynamicAccess$$Lambda$135/251210093.apply(Unknown Source)
[JVM-2] 	at scala.util.Try$.apply(Try.scala:209)
[JVM-2] 	at akka.actor.ReflectiveDynamicAccess.getClassFor(ReflectiveDynamicAccess.scala:20)
[JVM-2] 	at akka.remote.serialization.DaemonMsgCreateSerializer.deserialize(DaemonMsgCreateSerializer.scala:118)
[JVM-2] 	at akka.remote.serialization.DaemonMsgCreateSerializer.$anonfun$fromBinary$1(DaemonMsgCreateSerializer.scala:103)
[JVM-2] 	at akka.remote.serialization.DaemonMsgCreateSerializer$$Lambda$469/936208452.apply(Unknown Source)
```

For actors that are to be remote deployed the class based `Props` must be used:

```
Props(classOf[SomeActor], someConstructorParam)
```

**Comment by patriknw** [comment]
The generated lambdas are serializable if the requested SAM is. For scala.FunctionN, lambdas are always serializable. The problem we are seeing here is that the lambda is serialized as the top level object and for that we try to load the class with the generated lambda name. It works if I wrap the function in another Serializable class, e.g.

```scala
 case class Foo(fun: Function1[Int, Int])
```

We might be able to use that trick for the `Props(new SomeActor)` case also.

**Comment by JamesCarl** [comment]
Just wanted to say for those learning Akka this was a particularly tricky bug to identify when just getting started with remoting using Akka 2.4.16 and Scala version 2.12 a helpful note in the docs might increase adoption. Thank you for your work.

**Comment by patriknw** [comment]
Sorry for the inconvenience. It was a bug and it should be fixed in 2.4.17 so documentation should now be correct again. Let us know if you observe any thing else.

**Comment by JamesCarl** [comment]
Ah, just saw the new release thank you again!
